rust/src/records.rs

```diff
@@ -0,0 +1,507 @@
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        let mut p = Parser::new(value);
+        Ok(Self {
+            summary_start: p.get()?,
+            summary_offset_start: p.get()?,
+            summary_crc: p.get()?,
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Schema<'a> {
+    pub id: u16,
+    pub name: CowStr<'a>,
+    pub encoding: CowStr<'a>,
+    pub data: Cow<'a, [u8]>,
+}
+
+impl<'a> TryFrom<&'a [u8]> for Schema<'a> {
+    type Error = ParseError;
+    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
+        let mut p = Parser::new(value);
+        Ok(Self {
+            id: p.get()?,
+            name: p.get()?,
+            encoding: p.get()?,
+            data: p.get_byte_array()?,
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone)]
+pub struct Channel<'a> {
+    pub id: u16,
+    pub schema_id: u16,
+    pub topic: CowStr<'a>,
+    pub message_encoding: CowStr<'a>,
+    pub metadata: BTreeMap<CowStr<'a>, CowStr<'a>>,
+}
+
+impl<'a> TryFrom<&'a [u8]> for Channel<'a> {
+    type Error = ParseError;
+    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
+        let mut p = Parser::new(value);
+        Ok(Self {
+            id: p.get()?,
+            schema_id: p.get()?,
+            topic: p.get()?,
+            message_encoding: p.get()?,
+            metadata: p.get()?,
+        })
+    }
+}
+
+impl<'a> IntoStatic for Channel<'a> {
+    type Static = Channel<'static>;
+
+    fn into_static(self) -> Self::Static {
+        Channel {
+            id: self.id,
+            schema_id: self.schema_id,
+            topic: self.topic.into_static(),
+            message_encoding: self.message_encoding.into_static(),
+            metadata: self
+                .metadata
+                .iter()
+                .map(|(k, v)| (k.clone().into_static(), v.clone().into_static()))
+                .collect(),
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Message<'a> {
+    pub channel_id: u16,
+    pub sequence: u32,
+    pub log_time: Timestamp,
+    pub publish_time: Timestamp,
+    pub data: Cow<'a, [u8]>,
+}
+
+impl<'a> TryFrom<&'a [u8]> for Message<'a> {
+    type Error = ParseError;
+    fn try_from(value: &'a [u8]) -> Result<Self, Self::Error> {
+        let mut p = Parser::new(value);
+        Ok(Self {
+            channel_id: p.get()?,
+            sequence: p.get()?,
+            log_time: p.get()?,
+            publish_time: p.get()?,
+            data: Cow::Borrowed(p.into_inner()),
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Serialize, Deserialize, Clone, IntoStatic)]
+pub struct Chunk<'a> {
+    pub message_start_time: Timestamp,
+    pub message_end_time: Timestamp,
+    pub uncompressed_size: u64,
+    pub uncompressed_crc: u32,
+    pub compression: CowStr<'a>,
+    pub records: Cow<'a, [u8]>,
```

Comment:
How does using Cow differ from just `&'a [u8]`?

Reply:
Cow is an enum of `&'a [u8]` and `Vec<u8>` with some fancy sugar around it - using Cow will let someone construct a record by pushing to a `Vec<u8>` in the future. This will become more relevant when writing gets implemented.